pkg/apis/migration/v1alpha1/directimagemigration_types.go

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright 2020 Red Hat Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"strings"
+
+	imagev1 "github.com/openshift/api/image/v1"
+	kapi "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// DirectImageMigrationSpec defines the desired state of DirectImageMigration
+type DirectImageMigrationSpec struct {
+	SrcMigClusterRef  *kapi.ObjectReference `json:"srcMigClusterRef,omitempty"`
+	DestMigClusterRef *kapi.ObjectReference `json:"destMigClusterRef,omitempty"`
+	Namespaces        []string              `json:"namespaces,omitempty"`
+}
+
+// DirectImageMigrationStatus defines the observed state of DirectImageMigration
+type DirectImageMigrationStatus struct {
+	Conditions
+	ObservedDigest string       `json:"observedDigest,omitempty"`
+	StartTimestamp *metav1.Time `json:"startTimestamp,omitempty"`
+	Phase          string       `json:"phase,omitempty"`
+	Itinerary      string       `json:"itinerary,omitempty"`
+	Errors         []string     `json:"errors,omitempty"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DirectImageMigration is the Schema for the directimagemigrations API
+// +k8s:openapi-gen=true
+type DirectImageMigration struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   DirectImageMigrationSpec   `json:"spec,omitempty"`
+	Status DirectImageMigrationStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DirectImageMigrationList contains a list of DirectImageMigration
+type DirectImageMigrationList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []DirectImageMigration `json:"items"`
+}
+
+func (r *DirectImageMigration) GetSourceCluster(client k8sclient.Client) (*MigCluster, error) {
+	return GetCluster(client, r.Spec.SrcMigClusterRef)
+}
+
+func (r *DirectImageMigration) GetDestinationCluster(client k8sclient.Client) (*MigCluster, error) {
+	return GetCluster(client, r.Spec.DestMigClusterRef)
+}
+
+// GetSourceNamespaces get source namespaces without mapping
+func (r *DirectImageMigration) GetSourceNamespaces() []string {
+	includedNamespaces := []string{}
+	for _, namespace := range r.Spec.Namespaces {
+		namespace = strings.Split(namespace, ":")[0]
```

Comment:
is this because `:` is not allowed in the name of a namespace?

Reply:
It's because the namespace mapping is delimited by ":"
If the value is `["ns1", "ns2:ns3"]`, then imagestreams in `ns1` will be migrated to the same dest namespace, `ns1`, and imagestreams in `ns2` will be migrated to `ns3` -- this is the same format used for the namespace list on the migplan.

Reply:
I did not know that, TIL, thanks!